### PR TITLE
Improve ObjC metadata parsing and task-based parsing, increase performance

### PR DESCRIPTION
**Pull request by kabiroberai**
Firstly, thank you for making such a useful framework!

At the moment, this PR makes two main changes, as explained in the commit messages. I haven't fully finished integrating it yet, so I'm not sure whether I'll make more commits, which is why this is currently a draft. If you think it's ready to be merged, though, that's okay with me too (and if I make any future changes I can probably submit a separate PR).

There's also two `TODO`s that I added to `MKPtr.m`, but I'm not sure what the best way to go about either would be, so I'd appreciate your feedback on those.

### CODE REVIEW

**Comment by DeVaukz** [comment]
I am aware that various crash reporting libraries - some currently shipping in thousands of apps - use the mach_* APIs on embedded without any issues from Apple.  But I would prefer to keep the *stock* Mach-O Kit using strictly public API.  

How about this: Let's gate the mach_* API usage on iOS behind a new define, like MK_USE_MACH_API (name is open for bike-shedding).  By default this would be the value of MK_HAVE_MACH_VM.  Clients who are ok with using private API on embedded, could define it to 1 when building Mach-O Kit.  We can note this in the ReadMe.

**Comment by kabiroberai** [comment]
Another alternative I considered is using `vm_*` APIs (which _are_ public on iOS, as far as I can tell). A downside is that we'd need to drop the `mach_` prefix wherever it's used, though. I went for this approach since I felt it was simpler, but if you'd prefer to stick with public APIs then I think using vm_* is an option too

**Comment by kabiroberai** [comment]
Quick question related to another change I'm working on: would it be possible to support `MKPointer`s that point outside the Mach-O? I'm using MachO-Kit to parse images in-memory, and I currently have to work with the `memoryMap` directly in some cases, because pointers are outside the MachO due to modifications made to `__DATA` by the objc runtime (and also the dyld shared cache's uniquing in situations where system libraries are being parsed).

An example of the former is that the runtime replaces `class_ro_t` pointers (currently `MKObjCClassData`) inside objc classes with `class_rw_t` instances that have a heap copy of the `ro_t` counterparts. Atm I'm forced to entirely ditch `MKObjCClassData` and parse the `class_ro_t` myself because it can't find a corresponding node in the MachO image. Changing this would be a huge step towards better in-memory parsing capabilities – which I assume is in-scope, since `MKMachOImageProcessedByDYLD` is an option.

**Comment by DeVaukz** [comment]
Mach-O Kit does not use the `vm_*` types because they are defined based on the pointer size of the current process.  Admittedly, with Darwin now being 64-bits only this is no longer an issue.  However, I am seeing the `mach_vm_*` types defined in the latest iOS/tvOS SDK headers.  We can remove the `MK_HAVE_MACH_VM` checks and define the `mk_vm_*` types to their `mach_vm_*` counterparts on all platforms.

`_MKTaskMemoryMap` still can not build on embedded as-is - the `mach_vm_*` APIs are not defined in the headers.  Can the `vm_*` APIs handle safely mapping memory from another task?

**Comment by DeVaukz** [comment]
> Quick question related to another change I'm working on: would it be possible to support `MKPointer`s that point outside the Mach-O?

Yes.  This is something Mach-O Kit should add support for.  I'm not yet sure what the implementation should look like.  Any suggestions?

> An example of the former is that the runtime replaces `class_ro_t` pointers (currently `MKObjCClassData`) inside objc classes with `class_rw_t` instances that have a heap copy of the `ro_t` counterparts. Atm I'm forced to entirely ditch `MKObjCClassData` and parse the `class_ro_t` myself because it can't find a corresponding node in the MachO image. Changing this would be a huge step towards better in-memory parsing capabilities – which I assume is in-scope, since `MKMachOImageProcessedByDYLD` is an option.

I'm hesitant to add parsers for the runtime data structures to stock Mach-O Kit.  They are a private implementation detail, and Apple has been tweaking their layout heavily in the past couple releases.  Mach-O Kit should focus on making it easy for users who want to parse these structures to write their own MKNode subclasses.

File: MachOKit/Core/Impl/MKPtr.m [hunk 1]
```diff
@@ -66,6 +67,20 @@
...
+    // (slid) address of the pointee. MKPtr instead expects the address to follow
+    // MKNodeVMAddress semantics, so we must adjust for that.
+    // TODO - Add a test for this situation, and check if there are edge cases
+    // where dyld doesn't slide the pointer
+    MKMachOImage *image = [node nearestAncestorOfType:MKMachOImage.class];
+    if (image && image.isFromMemory) {
+        // TODO - For some reason, mk_vm_address_remove_slide thinks that
+        // `(mk_vm_address_t)(slide * -1) > addr` is true, so it returns
+        // MK_EUNDERFLOW. I'm not sure whether that's a bug here or in
+        // that function, but it should probably be looked into.
```
Comment (DeVaukz) [inline MachOKit/Core/Impl/MKPtr.m]: It is a bug in `mk_vm_address_remove_slide`.  Thanks!

**Comment by kabiroberai** [comment]
Hey! I've had a chance to work on this a little more, and I wanted to mention a few changes I made along with some thoughts I had. I'll probably push these changes to this branch soon but if you want me to move the changes I'm about to push into a separate PR then feel free to cherry-pick just the initial commits.

Firstly, regarding `mach_vm` vs `vm` APIs, I think they're mostly identical (except for the types' sizes) so using `vm_*` functions should be alright.

I also agree that deferring runtime data structure parsing to users is probably a good idea. I realized there are some runtime structures that can't even be parsed without calling runtime APIs, because these structures are clobbered in `__DATA` and moved into a `static` hash table inside a runtime function.

As for pointers outside the MachO, one solution I can think of is creating an `MKNode` subclass/protocol (say `MKOrphanableNode`) which indicates that the type doesn't require a parent, and then perhaps adding a check in `MKPtrPointee` to directly instantiate the class at that address if it's an `MKOrphanableNode`.

Speaking of pointers, I managed to speed up the class parsing I'm doing by 60x (down from 2 minutes to 2 seconds) by fine tuning various `childNodeOccupyingVMAddress` implementations. The most common change was to switch to binary search rather than linear search to find nodes, but I also had to change `MKDataSection`'s implementation to simply create a new node if `_children` doesn't contain one at that exact address. While the `MKDataSection` change might result in a _slight_ memory usage increase (not from what I've seen though), it was the single biggest performance win by far so I hope you're alright with including that upstream.

Additionally, I also made another huge performance improvement by persisting the `MKNode` subclass cache beyond the autoreleasepool. When I tried parsing the load commands of all loaded libraries in the host process, it initially took ~5 minutes, but persisting the cache reduced this to around 1 _second_. I feel like the performance benefit significantly outweighs the convenience afforded in the (IMO relatively niche) use case of dynamically adding MKNode subclasses after the subclass cache has been built. To still accommodate this use case, there could be an `+[MKNode flushSubclassCache]` method or possibly a class variable such as `MKNode.automaticallyReloadsSubclassCache`.

I've also fixed some stuff pertaining to ObjC data structure parsing. In doing so, I had to get MachO-Kit to parse some tagged pointers, so I added initializers `MKPointer[Node]` which take a `mask` parameter. I had to make these the designated initializers, which _might_ be source breaking in some cases since the non-mask initializers are now convenience initializers. Finally, I had to make one more potentially source-breaking change, where some properties in `MKObjCCategory` and `MKObjCClass` are now `MKPointerNode`s instead of `MKPointer`s. These are fields which can refer to external symbols, due to which their values in the binary are zeroed out. To find the external symbol, it's necessary to look for a rebind action corresponding to the pointer node's own `nodeVMAddress` (which wouldn't be accessible using a regular `MKPointer`). In fact it might just be worth adding a better way to do this, such as `MKRebindablePointerNode` which can give you a bind action corresponding to its `nodeVMAddress` if its `address == 0`.

File: MachOKit/Core/Impl/MKPtr.m (hunk 1 above)
Comment (kabiroberai) [reply MachOKit/Core/Impl/MKPtr.m]: Would it make sense to use `__builtin_*_overflow` for overflow-related arithmetic operations? Those might be more efficient, assuming Clang doesn't realise that you're performing an overflow check the manual way

Comment (DeVaukz) [reply MachOKit/Core/Impl/MKPtr.m]: Yes.  Mach-O Kit should switch to the clang built-ins.

File: MachOKit/Core/MKNode.m [hunk 2]
```diff
@@ -197,13 +197,6 @@ + (NSSet*)subclasses
         
         if (cache)
             objc_storeWeak(cache, retValue);
-        
-        // We autorelease the returned set in order to force our cached
-        // subclasses to be flushed when the autorelease pool next drains.
-        // We can't cache the subclasses for too long because our client may
-        // load additional bundles or frameworks at runtime which may augment
-        // the available subclasses.
-        [retValue autorelease];
```
Comment (DeVaukz) [inline MachOKit/Core/MKNode.m:206]: I'm fine with this, and I trust your performance numbers.  I do want a mechanism for flushing the cache(s) when a new framework/bundle is loaded.  I like your suggestion for a `+[MKNode flushSubclassCache]` method.  The caching storage will need to be re-worked: get rid of `-[MKNode _subclassesCache]` with `__weak` storage and have `MKNode` instead keep a dictionary of subclass => cache that can quickly be cleared on cache flush.

I can handle that change after this PR is merged.

File: libMachO/Core/core.h [hunk 3]
```diff
@@ -38,10 +38,36 @@
 #include "base.h"
 
 #include <mach/mach.h>
-#if TARGET_OS_MAC && !TARGET_OS_IPHONE
-#   include <mach/mach_vm.h>
-#   define MK_HAVE_MACH_VM 1
-#endif
+
+kern_return_t mach_vm_allocate
```
Comment (DeVaukz) [inline libMachO/Core/core.h]: I am against using the `mach_*` functions on embedded because they are not included in the public SDK headers.  Mach-O Kit can switch to using the `vm_` APIs if they provide the required functionality.  Or gate usage of the `mach_*` APIs behind a compile flag like I suggested in my original comment.

File: MachOKit/Core/Memory/MKMemoryMap.m [hunk 4]
```diff
@@ -130,15 +128,15 @@ - (vm_size_t)copyBytesAtOffset:(mk_vm_offset_t)offset fromAddress:(mk_vm_address
...
+        localError = [error retain];
         if (error)
             return;
         
         memcpy(buffer, (void*)address, (vm_size_t)MIN(length, (mk_vm_size_t)mappingLength));
         retValue = (vm_size_t)length;
     }];
     
-    MK_ERROR_OUT = localError;
+    MK_ERROR_OUT = [localError autorelease];
```
Comment (DeVaukz) [inline MachOKit/Core/Memory/MKMemoryMap.m:149]: Good find.  This change should be applied to all of the convenience methods in MKMemoryMap.m that pass the error from `-remapBytesAtOffset:fromAddress:length:requireFull:withHandler:` back to the caller.

File: MachOKit/Segments/MKMachO+Segments.m [hunk 5]
```diff
@@ -109,6 +109,7 @@ - (NSDictionary*)_segments
             }
         }
         
+        [MKBackedNode sortOptionalNodeArray:(NSMutableArray *)segments];
```
Comment (DeVaukz) [inline MachOKit/Segments/MKMachO+Segments.m]: I prefer to keep the 'All' segments array matching the order of the load commands.  

Create a copy of the segments array, sort it, and add it to the _segments dictionary under a new key (e.g, MKSortedSegments).  This will eliminate the need to handle sorting optionals - any optional without a value can be discarded from the sorted array.

File: MachOKit/Segments/MKSegment.m [hunk 6]
```diff
@@ -198,6 +198,7 @@ - (instancetype)initWithLoadCommand:(id<MKLCSegment>)segmentLoadCommand error:(N
             }
         }
         
+        [MKBackedNode sortOptionalNodeArray:(NSMutableArray *)segmentSections];
```
Comment (DeVaukz) [inline MachOKit/Segments/MKSegment.m]: Like the segments array, I prefer to keep the sections array matching the order of the load command.

File: MachOKit/DataInCode/MKDataInCode.m [hunk 7]
```diff
@@ -113,16 +113,9 @@ - (void)dealloc
...
-        mk_vm_range_t range = mk_vm_range_make(entry.nodeVMAddress, entry.nodeSize);
-        if (mk_vm_range_contains_address(range, 0, address) == MK_ESUCCESS) {
-            MKOptional *child = [entry childNodeOccupyingVMAddress:address targetClass:targetClass];
-            if (child.value)
-                return child;
-            // else, fallthrough and call the super's implementation.
-            // The caller may actually be looking for *this* node.
-        }
-    }
+    MKOptional *child = [MKBackedNode childNodeOccupyingVMAddress:address targetClass:targetClass inSortedArray:(NSArray *)self.entries];
```
Comment (DeVaukz) [inline MachOKit/DataInCode/MKDataInCode.m:116]: This creates an implicit requirement that `entries` is sorted in address order.  A unit test to verify this should be added, to prevent subtle bugs from later slipping in.  Same for all the other nodes with the change.

I can handle that after the PR is merged.

File: MachOKit/Segments/MKMachO+Segments.m (hunk 5 above)
Comment (kabiroberai) [reply MachOKit/Segments/MKMachO+Segments.m]: Done in bcac3dc

File: MachOKit/Segments/MKSegment.m (hunk 6 above)
Comment (kabiroberai) [reply MachOKit/Segments/MKSegment.m]: Handled in bcac3dc, similarly to the change in `MKMachO+Segments.m`

File: MachOKit/Core/Memory/MKMemoryMap.m (hunk 4 above)
Comment (kabiroberai) [reply MachOKit/Core/Memory/MKMemoryMap.m:149]: Fixed in a106c71

File: libMachO/Core/core.h (hunk 3 above)
Comment (kabiroberai) [reply libMachO/Core/core.h]: Would you be fine with switching to `vm_*` APIs on macOS as well? It might be messy to use both APIs internally since there would need to be preprocessor checks to determine which API to use, and additionally the types are slightly different. The outward facing types will still be `mach_` based as the types themselves are defined even on iOS.

Comment (kabiroberai) [reply libMachO/Core/core.h]: FWIW this change turned out to be easier than expected and I've implemented it in 430a952. If you'd rather use `mach_vm_*` on macOS, I can add that back with a preprocessor check for the macOS platform.

File: MachOKit/Segments/Sections/MKDataSection.m [hunk 8]
```diff
@@ -65,20 +65,17 @@ - (void)dealloc
...
-    if (child)
-        return [child autorelease];
-    else
-        return [super childNodeOccupyingVMAddress:address targetClass:targetClass];
+    // While we could also check whether `address` is _within_ any children, a dict lookup results
+    // in a *net* 60x speed-up, and in most cases `address` directly corresponds to a child anyway.
+    // It might be possible to increase efficiency of the other lookup method using a trie or a
+    // binary search tree but that will probably have a negligible memory improvement and a measurable
+    // impact on performance.
+
```
Comment (DeVaukz) [inline MachOKit/Segments/Sections/MKDataSection.m]: This method should return the correct child if 'address' is *within* a child node.

I suggest that if _children[@(address)] is nil the implementation checks if 'address' is within *this* (self) node and if it is, falls back to the previous linear search.  That will keep the performance gains for the common case while maintaining the existing semantics for the edge cases.

File: MachOKit/Shared/MKPointer+Node.m [hunk 9]
```diff
@@ -58,7 +64,13 @@ - (instancetype)initWithOffset:(mk_vm_offset_t)offset fromParent:(MKBackedNode*)
...
     }
     
-    return [self initWithOffset:offset fromParent:parent context:context error:error];
+    return [self initWithOffset:offset fromParent:parent mask:mask context:context error:error];
+}
+
+//|++++++++++++++++++++++++++++++++++++|//
+- (instancetype)initWithOffset:(mk_vm_offset_t)offset fromParent:(MKBackedNode*)parent targetClass:(Class)targetClass error:(NSError**)error
+{
+    return [self initWithOffset:offset fromParent:parent mask:(mk_vm_address_t)(~0) targetClass:targetClass error:error];
```
Comment (DeVaukz) [inline MachOKit/Shared/MKPointer+Node.m]: Please make the mask parameter here consistent with the one in `initWithOffset:fromParent:context:error:` by casting to `mk_vm_address_t` before the Binary NOT operation.

**Comment by DeVaukz** [comment]
> Finally, I had to make one more potentially source-breaking change, where some properties in MKObjCCategory and MKObjCClass are now MKPointerNodes instead of MKPointers. These are fields which can refer to external symbols, due to which their values in the binary are zeroed out. To find the external symbol, it's necessary to look for a rebind action corresponding to the pointer node's own nodeVMAddress (which wouldn't be accessible using a regular MKPointer). In fact it might just be worth adding a better way to do this, such as MKRebindablePointerNode which can give you a bind action corresponding to its nodeVMAddress if its address == 0.

This Is a longstanding issue that needs more encompassing solution.  In theory, any pointer inside of a Mach-O could be external.  MKPointerNode is not the right type to handle this.  It is intended to represent a node with a pointer as the only field (e.g, the pointers in a `S_LITERAL_POINTERS` section), not a field in some other node.

I will merge the current change, but I have an idea to solve this issue using MKPointer.

File: MachOKit/ObjC/MKObjCElementList.m [hunk 10]
```diff
@@ -155,7 +155,7 @@ + (Class)classForGenericArgumentAtIndex:(__unused NSUInteger)index
 
 //|++++++++++++++++++++++++++++++++++++|//
 + (uint32_t)flagsMask
-{ return 0; }
+{ return 3; }
```
Comment (DeVaukz) [inline MachOKit/ObjC/MKObjCElementList.m]: What is the reason for this change?  According to the ObjC runtime source, only `method_list_t` reserves the least significant two bits for flags.  `MKObjCClassMethodList` should override `+flagsMask` to return 3 instead of changing the default implementation.

https://opensource.apple.com/source/objc4/objc4-781/runtime/objc-runtime-new.h.auto.html

**Comment by DeVaukz** [comment]
PR is looking good.  Once the last three issues are resolved, I will merge it.

File: MachOKit/Segments/Sections/MKDataSection.m (hunk 8 above)
Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: I just spent a while looking into this, and I realised there's a bigger problem with `MKDataSection`: the semantics of `-[MKDataSection childNodeOccupyingVMAddress:]` were inherently non-deterministic. As I understand it, the point of `childNodeOccupyingVMAddress` is to find a suitable parent for the child node being created. However, since `MKDataSection` lazily creates child nodes, this results in a problem:

To illustrate this issue, let's say we have an `MKDataSection` which contains nodes of type `MKNodeA` (nodeSize=16), which in turn has a node of type `MKNodeB` (nodeSize=8) at offset 8. The issue is that if we dereference an `MKPointer<MKNodeB>`, it _may or may not_ have `MKNodeA` as its parent depending on whether we've previously dereferenced the `MKNodeA` which "contains" it. This is because `-[MKPointer<MKNodeB> pointee]` would call `-[MKDataSection childNodeAtVMAddress:foo]`, which would create the `MKNodeB` and then call `-[MKDataSection childNodeOccupyingVMAddress:foo]`. Now, if the `MKNodeA` at `[foo - 8, foo + 8)` had been dereferenced previously, the linear search through `_children` would locate it and return it. However, if it had not been created in the past, the linear search would fail and the parent would be set to the `MKDataSection` itself. Thus the expected semantics of `childNodeOccupyingVMAddress` were not being followed in the first place (or at least not always).

In light of this, I realize that my change would make things _more_ deterministic by making it so that an element in an `MKDataSection` always has its parent set to the data section, rather than it depending on whether or not the "real" parent had previously been dereferenced. As far as I can tell, the only way to have determinisim while also maintaining the actual hierarchy would be to non-lazily parse the data section or do some sort of look-behind, however I think it's better to view `MKDataSection` as an unstructured blob because the alternative would require knowledge of the exact format of the data section, and I'm not too sure whether that's well-defined. For example, if an `MKDataSection` is heterogeneous then I don't think it'd even be possible to non-lazily parse it in a way that isn't super hacky.

I did try implementing your suggestion though, and unfortunately the performance is about the same as what it was with a plain linear search, as there are a lot of pointers that point into an `MKDataSection` while parsing classes. It might be possible to overcome this by using an AVL tree to store `_children`, but that would result in a degradation from O(1) to O(log n) performance (in `childNodeOccupyingVMAddress` due to tree searching, and `childNodeAtVMAddress` due to tree insertion), and would additionally retain the non-deterministic behavior of the node's `parent` depending on prior dereferences.

I think the best solution would be to make `MKDataSection`'s children always have the data section as its parent – which I unknowingly already did in my PR, however I can now formalize that. This would result in deterministic behavior plus O(1) lookup time. In fact if this is the defined behavior, another optimization will be possible in that `-[MKDataSection childNodeAtVMAddress:]` can create `targetClass` using `initWithOffset:fromParent:`, skipping the extra step in `initWithVMAddress` to find the parent (as that would, by definition, return the data section anyway).

File: MachOKit/Shared/MKPointer+Node.m (hunk 9 above)
Comment (kabiroberai) [reply MachOKit/Shared/MKPointer+Node.m]: Fixed in 5a86727

File: MachOKit/ObjC/MKObjCElementList.m (hunk 10 above)
Comment (kabiroberai) [reply MachOKit/ObjC/MKObjCElementList.m]: My bad, I've corrected this in fd20729.

File: MachOKit/Segments/Sections/MKDataSection.m (hunk 8 above)
Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: Btw I've committed the aforementioned changes in 583bcf2 in case you'd like to have a look, however I can revert them if necessary.

Comment (DeVaukz) [reply MachOKit/Segments/Sections/MKDataSection.m]: That MKDataSection creates child nodes on-demand is not ideal but is the consequence of there being no way to parse the contents of a data section without information from elsewhere, as you already figured out.

Still thinking about the rest of your comment...

Comment (DeVaukz) [reply MachOKit/Segments/Sections/MKDataSection.m]: `-childNodeOccupyingVMAddress:targetClass:` should maintain the semantics of always returning the *deepest* node in the tree that occupies *address* and is a *targetClass*.  With your current implementation, a client would not be able to say, find the `MKObjCClassMethod` occupying an address because `MKObjCClassMethod` nodes are children of a `MKObjCClassMethodList` node, which is a child of a `MKDataSection` [1].  I understand that this will mean keeping the existing linear search (for now).

`-childNodeAtVMAddress:targetClass:` should follow the same *find the deepest node* semantics of -childNodeOccupyingVMAddress:targetClass:.  My current implementation of `-[MKDataSection childNodeAtVMAddress:targetClass:]` does not.  That and the other flaws you pointed out have led me to conclude overloading `-[MKDataSection childNodeAtVMAddress:targetClass:]` to lazily create child nodes was a mistake.

The current implementation for resolving pointers and lazily creating the pointee nodes should be reworked - possibly by moving to a new method which can be given more suitable and performant semantics.  That can be done later - I will open an issue with more thoughts.  Your changes to `-[MKDataSection childNodeAtVMAddress:targetClass:]` are fine for now.

Reverting the changes to `-[MKDataNode childNodeOccupyingVMAddress:targetClass:]` is the only remaining change needed to this PR.

[1] Or the `MKDataSection` would create the `MKObjCClassMethod` node and add it as a child, which I do not want.  An `MKObjCClassMethod` should be a child of the `MKObjCClassMethodList` that it resides within.

Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: Hmm I'm not entirely sure I understand the purpose of reverting the change to `childNodeOccupyingVMAddress`. As far as I can tell, the previous implementation of the aforementioned method had the same flaw as `childNodeAtVMAddress`, in that it was non-deterministic. Even with the change reverted, the semantics would only be correct in the case where the `MKObjCClassMethodList` had already been dereferenced in the past. Otherwise, it would still create the `MKObjCClassMethod` with the `MKDataSection` as its parent.

IMO it would make more sense to revert `childNodeOccupyingVMAddress` once the implementation of `childNodeAtVMAddress` has been changed as you stated, because at that point it would probably also be possible to use a better data structure to remove the need for a linear search (for example, if you decide to non-lazily parse the section somehow, it would make sense to use a sorted array). Reverting this atm would probably not make any improvements to "correctness" (since the method would continue to be non-deterministic), however it would degrade performance significantly. Additionally, I don't think this affects any real-world use cases either, because from what I've seen there are no nodes within `MKDataSection` that have such a hierarchy – for example in the case of `MKObjCClassMethodList`, the element of the list is a pointer (or relative offset on iOS 14) to the method, and not the method itself, so the method would never have the list as its parent anyway.

Comment (DeVaukz) [reply MachOKit/Segments/Sections/MKDataSection.m]: The elements of MKObjCClassMethodList are not pointers.  The runtime defines `method_list_t` as an `entsize_list_tt` of `method_t` structures.

`struct method_list_t : entsize_list_tt<method_t, method_list_t, 0x3>`

Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: Oops, I was looking at the definitions of the list sections instead. For the `entsize_list`s, though, aren't the elements explicitly created with parents anyway? https://github.com/DeVaukz/MachO-Kit/blob/20f5fe2645e2707689135e5ee334a678fd947737/MachOKit/ObjC/MKObjCElementList.m#L102

And in case there's a pointer to a method inside a method list from elsewhere, even with my change reverted, whether or not the `MKObjCClassMethod` has the list as its parent will depend on whether the method list has previously been dereferenced.

Comment (DeVaukz) [reply MachOKit/Segments/Sections/MKDataSection.m]: The problem is that with your implementation of `-[MKDataSection childNodeOccupyingVMAddress:targetClass:]`, this code returns the data section instead of the method:

```
let objCClass: MKObjCClass = // Assume this points to a class with at least one method
let objCClassData = objCClass.classData.pointee.value!
let methodList = objCClassData.methods.pointee.value!
let method = methodList.elements.first!
let methodAddress = method.nodeVMAddress
    
let foo = macho.childNodeOccupyingVMAddress(methodAddress, targetClass: nil)
print(foo)
```

With this change:
`__DATA,__objc_const`

With existing implementation:
`<MKObjCClassMethod 0x7ffa6add0b80: address = 0x6a988, size = 0x18>`

Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: Ah, I see; would you be okay with cherry-picking to omit that particular change then? I'd like to keep it downstream since I don't need arbitrary access to pointers within `MKDataSection` myself, but I understand why you wouldn't want it upstream for the sake of correctness. 

I'm about to push one more commit which fixes a few memory leaks, but this should be ready to merge once that's done. Also, I haven't had a chance to run all the tests, so you might also want to run those first in case you haven't yet (speaking of which, it's probably worth setting up a CI for that)

Comment (kabiroberai) [reply MachOKit/Segments/Sections/MKDataSection.m]: Or actually nevermind that, I'll just move my `childNodeOccupyingVMAddress` change into another branch so that master is ready to merge without needing to cherry-pick

Comment (DeVaukz) [reply MachOKit/Segments/Sections/MKDataSection.m]: Looks good.  Merging.

**Comment by DeVaukz** [comment]
Thank you!
